kotlin/lib/src/main/kotlin/truelayer/signing/Jws.kt

```diff
@@ -0,0 +1,90 @@
+package truelayer.signing
+
+import com.nimbusds.jose.JWSAlgorithm
+import com.nimbusds.jose.JWSHeader
+import com.nimbusds.jose.JWSObject
+import com.nimbusds.jose.Payload
+import com.nimbusds.jose.crypto.ECDSAVerifier
+import com.nimbusds.jose.util.Base64URL
+import java.security.interfaces.ECPublicKey
+
+internal fun validateSignature(
+    signature: String,
+    signatureHeaderNames: List<String>,
+    headers: Map<HeaderName, String>,
+    method: String,
+    path: String,
+    body: ByteArray,
+    publicKey: ECPublicKey
+): Result<Unit> {
+    val orderedHeaders =
+        signatureHeaderNames.fold(LinkedHashMap<HeaderName, String>(0)) { acc, curr ->
+            val value: String? = headers[HeaderName((curr))]
+            if (value != null) {
+                acc[HeaderName(curr)] = value
+            }
+            acc
+        }
+
+    return InvalidSignatureException.evaluate {
+        JWSObject.parse(signature, Payload(Base64URL(buildPayload(orderedHeaders, method, path, body))))
+            .verify(ECDSAVerifier(publicKey))
+    }.flatMap { verified -> InvalidSignatureException.ensure({ verified }, "jws body mismatch") }
```

Comment:
This is misleading, kinda suggests that the http body is different. I'd stick with whats used in other langs at this point `Invalid signature`. The signature is wrong, and thats pretty much all we can say.

```suggestion
    }.flatMap { verified -> InvalidSignatureException.ensure({ verified }, "invalid signature") }
```